server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java

```diff
@@ -0,0 +1,271 @@
+            return;
+        }
+
+        DiscoveryNode healthNode = HealthNode.findHealthNode(event.state());
+        if (healthNode == null) {
+            this.isHealthNode = false;
+            this.maybeCancelJob();
+            return;
+        }
+        final boolean isCurrentlyHealthNode = healthNode.getId().equals(this.clusterService.localNode().getId());
+        if (this.isHealthNode != isCurrentlyHealthNode) {
+            this.isHealthNode = isCurrentlyHealthNode;
+            if (this.isHealthNode) {
+                // we weren't the health node, and now we are
+                maybeScheduleJob();
+            } else {
+                // we were the health node, and now we aren't
+                maybeCancelJob();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        SchedulerEngine engine = scheduler.get();
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(HEALTH_PERIODIC_LOGGER_JOB_NAME)) {
+            this.tryToLogHealth();
+        }
+    }
+
+    // default visibility for testing purposes
+    void tryToLogHealth() {
+        if (this.currentlyRunning.compareAndExchange(false, true) == false) {
+            // We ensure that the flag will only be released once when it's time
+            RunOnce release = new RunOnce(() -> currentlyRunning.set(false));
+            try {
+                // We can use the runAfter listener to wire the release to the current listener
+                ActionListener<List<HealthIndicatorResult>> listenerWithRelease = ActionListener.runAfter(resultsListener, release);
+                this.healthService.getHealth(this.client, null, false, 0, listenerWithRelease);
+            } catch (Exception e) {
+                logger.warn(() -> "The health periodic logger encountered an error.", e);
+                // In case of an exception before the listener was wired, we can release the flag here, and we feel safe
+                // that it will not release it again because this can only be run once.
+                release.run();
+            }
+        }
+    }
+
+    // default visibility for testing purposes
+    SchedulerEngine getScheduler() {
+        return this.scheduler.get();
+    }
+
+    /**
+     * Create a Map of the results, which is then turned into JSON for logging.
+     *
+     * The structure looks like:
+     * {"elasticsearch.health.overall.status": "green", "elasticsearch.health.[other indicators].status": "green"}
+     * Only the indicator status values are included, along with the computed top-level status.
+     *
+     * @param indicatorResults the results of the Health API call that will be used as the output.
+     */
+    // default visibility for testing purposes
+    Map<String, Object> toLoggedFields(List<HealthIndicatorResult> indicatorResults) {
```

Comment:
If you make this method static you would be able to use it (and test it) without instantiating the whole `HealthPeriodicLogger`. On  more careful look, I would rename it to `convertToLoggedFields`. This might be just me, but reading a method `healthPeriodicLogger.toLoggedFields` my first assumption would be that this converts the instance of `healthPeriodicLogger` to logged fields. While prepending with `convert` and making it static makes it a bit more explicit that the input is the argument not the calling object.

Does this make sense?